text/0069-schema-libraries.md

```diff
@@ -0,0 +1,169 @@
+# Schema libraries
+
+## Related issues and PRs
+
+- Reference Issues: Inspired by [RFC 58], but includes new material not present in RFC 58
+- Implementation PR(s):
+
+## Timeline
+
+- Started: 2024-06-07
+
+## Summary
+
+Allow Cedar schemas to include/import "libraries" of definitions from remote
+URLs.
+
+This RFC does not propose that the Cedar team would build or maintain any such
+libraries; it only proposes the mechanism for importing libraries from URLs.
+
+## Basic example
+
+Human schema format:
+```
+import "https://raw.githubusercontent.com/cedar-policy/cedar-examples/release/3.2.x/cedar-example-use-cases/document_cloud/document_cloud.cedarschema"
+import "https://example.com/cedar_schemas/oidc.cedarschema"
+import "https://example.com/cedar_schemas_json/foobar.cedarschema.json"
+
+namespace "MyApp" {
+  ...
+}
+```
+
+JSON schema format:
+```
+{
+  "imports" : [
+    "https://raw.githubusercontent.com/cedar-policy/cedar-examples/release/3.2.x/cedar-example-use-cases/document_cloud/document_cloud.cedarschema",
+    "https://example.com/cedar_schemas/oidc.cedarschema",
+    "https://example.com/cedar_schemas_json/foobar.cedarschema.json"
+  ],
+  "MyApp": {
+    ...
+  }
+}
+```
+
+## Motivation
+
+Some data sources are common across many applications and useful to many Cedar
+users, either within the same organization or even across organizations.
+
+### 1. Within the same organization
+
+This RFC would allow an organization to define its own libraries of schema
+definitions, which could be reused across many different schemas (say, for
+different webapps owned by the organization).
+For instance, the organization might have common definitions of `User` or
+`Account` that apply in many different applications, and although those
+applications may not want to share entire Cedar schemas, with this RFC they
+could share just the definitions of `User` or `Account`, which could be defined
+once in a central location (in the same or separate libraries).
+
+### 2. Across organizations
+
+For another motivating example, consider identity providers (IdPs) which comply
+with the OpenID Connect standard (OIDC).
+The OIDC standard includes a list of attributes that exist on a user type; this
+is naturally declared as a Cedar entity type.
+With this RFC, anyone could provide a "library" representing Cedar definitions for
+OIDC types, and provide that library as a Cedar schema file at some URL; and then
+other Cedar users could use those definitions simply by importing the file from that URL.
+This would allow the Cedar community to gradually coalesce on the "best" way to
+represent an OIDC user in Cedar.
+
+### Motivations common to both scenarios
```

Comment:
Could we also have mechanism to `use` a particular namespace? For instance, say I'm building an app where everything is under the AWS namespace. It would be nice if I could just `use AWS::IdentityCenter` and then policy authors only need to specify the sub-entity. e.g. `User` vs `AWS::IdentityCenter::User`.

Reply:
I can totally see this being useful, but I'm torn whether it should be part of this RFC or a separate RFC.

It would be harder to envision what a `use` would look like for policies (as opposed to schemas); I'm not sure if users will be happy with a `use` mechanism for schemas but not having one for policies?

Reply:
Yeah, its most useful for policies IMO. Perhaps a separate RFC is better.